File: apps/watchlist/backend/watchlist_app/services/return_windows.py
from __future__ import annotations

from calendar import monthrange
from dataclasses import dataclass
from datetime import date, timedelta
from math import isfinite
from typing import Any, Literal


RETURN_WINDOW_POLICY_VERSION = "return-window/v1"
AnchorMode = Literal["on_or_before", "strictly_before"]


@dataclass(frozen=True)
class ReturnWindow:
    requested_start_date: date
    requested_end_date: date
    anchor_mode: AnchorMode
    anchor_date: date
    end_date: date
    points: tuple[dict[str, Any], ...]

    @property
    def elapsed_days(self) -> int:
        return max((self.end_date - self.anchor_date).days, 0)


@dataclass(frozen=True)
class ReturnWindowSpec:
    requested_start_date: date
    requested_end_date: date
    anchor_mode: AnchorMode


def _shift_months(value: date, months: int) -> date:
    month_index = value.year * 12 + value.month - 1 + months
    target_year, target_month_index = divmod(month_index, 12)
    target_month = target_month_index + 1
    target_day = min(value.day, monthrange(target_year, target_month)[1])
    return date(target_year, target_month, target_day)


def named_return_window_spec(window: str, as_of_date: date) -> ReturnWindowSpec:
    normalized = window.strip().upper()
    if normalized == "1D":
        start = as_of_date - timedelta(days=1)
        mode: AnchorMode = "on_or_before"
    elif normalized == "1W":
        start = as_of_date - timedelta(days=7)
        mode = "on_or_before"
    elif normalized == "1M":
        start = _shift_months(as_of_date, -1)
        mode = "on_or_before"
    elif normalized == "3M":
        start = _shift_months(as_of_date, -3)
        mode = "on_or_before"
    elif normalized == "6M":
        start = _shift_months(as_of_date, -6)
        mode = "on_or_before"
    elif normalized in {"1Y", "2Y", "3Y", "5Y", "10Y"}:
        start = _shift_months(as_of_date, -12 * int(normalized[:-1]))
        mode = "on_or_before"
    elif normalized == "MTD":
        start = date(as_of_date.year, as_of_date.month, 1)
        mode = "strictly_before"
    elif normalized == "YTD":
        start = date(as_of_date.year, 1, 1)
        mode = "strictly_before"
    else:
        raise ValueError(f'Unsupported return window "{window}".')
    return ReturnWindowSpec(
        requested_start_date=start,
        requested_end_date=as_of_date,
        anchor_mode=mode,
    )


def resolve_return_window(
    points: list[dict[str, Any]],
    *,
    requested_start_date: date,
    requested_end_date: date,
    anchor_mode: AnchorMode = "on_or_before",
) -> ReturnWindow | None:
    if requested_start_date > requested_end_date:
        return None

    normalized_by_date: dict[date, dict[str, Any]] = {}
    for raw_point in points:
        point_date = raw_point.get("as_of_date")
        raw_value = raw_point.get("value")
        if not isinstance(point_date, date) or isinstance(raw_value, bool):
            continue
        try:
            value = float(raw_value)
        except (TypeError, ValueError):
            continue
        if not isfinite(value) or value <= 0:
            continue
        normalized_by_date[point_date] = {**raw_point, "value": value}

    ordered = [normalized_by_date[key] for key in sorted(normalized_by_date)]
    end_index = next(
        (
            index
            for index in range(len(ordered) - 1, -1, -1)
            if ordered[index]["as_of_date"] <= requested_end_date
        ),
        -1,
    )
    if end_index < 0:
        return None

    def is_anchor_candidate(point: dict[str, Any]) -> bool:
        point_date = point["as_of_date"]
        if anchor_mode == "strictly_before":
            return point_date < requested_start_date
        return point_date <= requested_start_date

    anchor_index = next(
        (
            index
            for index in range(end_index, -1, -1)
            if is_anchor_candidate(ordered[index])
        ),
        -1,
    )
    if anchor_index < 0 or anchor_index >= end_index:
        return None

    window_points = tuple(ordered[anchor_index : end_index + 1])
    return ReturnWindow(
        requested_start_date=requested_start_date,
        requested_end_date=requested_end_date,
        anchor_mode=anchor_mode,
        anchor_date=window_points[0]["as_of_date"],
        end_date=window_points[-1]["as_of_date"],
        points=window_points,
    )


def period_return_percent(window: ReturnWindow) -> float:
    first_value = float(window.points[0]["value"])
    last_value = float(window.points[-1]["value"])
    return (last_value / first_value - 1) * 100


def annualized_return_percent(window: ReturnWindow) -> float | None:
    if window.elapsed_days <= 0:
        return None
    first_value = float(window.points[0]["value"])
    last_value = float(window.points[-1]["value"])
    return (pow(last_value / first_value, 365.25 / window.elapsed_days) - 1) * 100


def normalized_return_points(window: ReturnWindow) -> list[dict[str, object]]:
    first_value = float(window.points[0]["value"])
    return [
        {
            "date": point["as_of_date"].isoformat(),
            "value": (float(point["value"]) / first_value - 1) * 100,
        }
        for point in window.points
    ]


def return_window_metadata(window: ReturnWindow) -> dict[str, object]:
    return {
        "policy_version": RETURN_WINDOW_POLICY_VERSION,
        "anchor_mode": window.anchor_mode,
        "requested_start_date": window.requested_start_date.isoformat(),
        "requested_end_date": window.requested_end_date.isoformat(),
        "anchor_date": window.anchor_date.isoformat(),
        "end_date": window.end_date.isoformat(),
    }

File: apps/watchlist/frontend/src/lib/returnWindows.ts
import type { InstrumentChartPoint } from './api'

export const RETURN_WINDOW_POLICY_VERSION = 'return-window/v2'

export type ReturnAnchorMode = 'on_or_before' | 'strictly_before'

export type ReturnWindow = {
  requestedStartDate: string
  requestedEndDate: string
  anchorMode: ReturnAnchorMode
  anchorDate: string
  endDate: string
  points: InstrumentChartPoint[]
}

export type AlignedReturnWindows = {
  left: ReturnWindow
  right: ReturnWindow
}

export type ReturnWindowName =
  | '1D'
  | '1W'
  | '1M'
  | '3M'
  | '6M'
  | 'MTD'
  | 'YTD'
  | '1Y'
  | '2Y'
  | '3Y'
  | '5Y'
  | '10Y'

export function shiftIsoDate(
  value: string,
  offset: { years?: number; months?: number; days?: number },
) {
  const [year, month, day] = value.slice(0, 10).split('-').map(Number)
  if (![year, month, day].every(Number.isFinite)) {
    return null
  }
  const targetMonthIndex = month - 1 + (offset.months || 0) + (offset.years || 0) * 12
  const targetYear = year + Math.floor(targetMonthIndex / 12)
  const normalizedMonthIndex = ((targetMonthIndex % 12) + 12) % 12
  const lastDayOfMonth = new Date(Date.UTC(targetYear, normalizedMonthIndex + 1, 0)).getUTCDate()
  const targetDay = Math.min(day, lastDayOfMonth)
  const shifted = new Date(Date.UTC(targetYear, normalizedMonthIndex, targetDay))
  if (offset.days) {
    shifted.setUTCDate(shifted.getUTCDate() + offset.days)
  }
  return shifted.toISOString().slice(0, 10)
}

export function namedReturnWindowSpec(window: ReturnWindowName, asOfDate: string) {
  const end = asOfDate.slice(0, 10)
  if (window === 'MTD') {
    return { start: `${end.slice(0, 7)}-01`, end, anchorMode: 'strictly_before' as const }
  }
  if (window === 'YTD') {
    return { start: `${end.slice(0, 4)}-01-01`, end, anchorMode: 'strictly_before' as const }
  }
  const start =
    window === '1D'
      ? shiftIsoDate(end, { days: -1 })
      : window === '1W'
        ? shiftIsoDate(end, { days: -7 })
        : window.endsWith('M')
          ? shiftIsoDate(end, { months: -Number(window.slice(0, -1)) })
          : shiftIsoDate(end, { years: -Number(window.slice(0, -1)) })
  return { start: start || '', end, anchorMode: 'on_or_before' as const }
}

function normalizedReturnPoints(points: InstrumentChartPoint[]) {
  const byDate = new Map<string, InstrumentChartPoint>()
  points.forEach((point) => {
    const value = Number(point.value)
    const pointDate = point.date.slice(0, 10)
    if (pointDate && Number.isFinite(value) && value > 0) {
      byDate.set(pointDate, { date: pointDate, value })
    }
  })
  return [...byDate.values()].sort((left, right) => left.date.localeCompare(right.date))
}

export function resolveReturnWindow(
  points: InstrumentChartPoint[],
  requestedStartDate: string,
  requestedEndDate: string,
  anchorMode: ReturnAnchorMode = 'on_or_before',
): ReturnWindow | null {
  if (!requestedStartDate || !requestedEndDate || requestedStartDate > requestedEndDate) {
    return null
  }
  const ordered = normalizedReturnPoints(points)
  let endIndex = -1
  for (let index = ordered.length - 1; index >= 0; index -= 1) {
    if (ordered[index].date <= requestedEndDate) {
      endIndex = index
      break
    }
  }
  if (endIndex < 0) {
    return null
  }
  let anchorIndex = -1
  for (let index = endIndex; index >= 0; index -= 1) {
    const dateMatches =
      anchorMode === 'strictly_before'
        ? ordered[index].date < requestedStartDate
        : ordered[index].date <= requestedStartDate
    if (dateMatches) {
      anchorIndex = index
      break
    }
  }
  if (anchorIndex < 0 || anchorIndex >= endIndex) {
    return null
  }
  const windowPoints = ordered.slice(anchorIndex, endIndex + 1)
  return {
    requestedStartDate,
    requestedEndDate,
    anchorMode,
    anchorDate: windowPoints[0].date,
    endDate: windowPoints[windowPoints.length - 1].date,
    points: windowPoints,
  }
}

export function resolveAlignedReturnWindows(
  leftPoints: InstrumentChartPoint[],
  rightPoints: InstrumentChartPoint[],
  requestedStartDate: string | null,
  requestedEndDate: string,
  anchorMode: ReturnAnchorMode = 'on_or_before',
): AlignedReturnWindows | null {
  if (
    !requestedEndDate ||
    (requestedStartDate != null &&
      (!requestedStartDate || requestedStartDate > requestedEndDate))
  ) {
    return null
  }
  const leftByDate = new Map(
    normalizedReturnPoints(leftPoints).map((point) => [point.date, point]),
  )
  const rightByDate = new Map(
    normalizedReturnPoints(rightPoints).map((point) => [point.date, point]),
  )
  const commonDates = [...leftByDate.keys()]
    .filter((pointDate) => rightByDate.has(pointDate) && pointDate <= requestedEndDate)
    .sort()
  if (commonDates.length < 2) {
    return null
  }
  const endDate = commonDates[commonDates.length - 1]
  let anchorDate: string | null = requestedStartDate == null ? commonDates[0] : null
  if (requestedStartDate != null) {
    for (let index = commonDates.length - 1; index >= 0; index -= 1) {
      const pointDate = commonDates[index]
      const matches =
        anchorMode === 'strictly_before'
          ? pointDate < requestedStartDate
          : pointDate <= requestedStartDate
      if (matches) {
        anchorDate = pointDate
        break
      }
    }
  }
  if (anchorDate == null || anchorDate >= endDate) {
    return null
  }
  const alignedDates = commonDates.filter(
    (pointDate) => pointDate >= anchorDate && pointDate <= endDate,
  )
  if (alignedDates.length < 2) {
    return null
  }
  const resolvedRequestedStartDate = requestedStartDate ?? anchorDate
  const buildWindow = (
    pointsByDate: Map<string, InstrumentChartPoint>,
  ): ReturnWindow => ({
    requestedStartDate: resolvedRequestedStartDate,
    requestedEndDate,
    anchorMode,
    anchorDate,
    endDate,
    points: alignedDates.map((pointDate) => pointsByDate.get(pointDate)!),
  })
  return {
    left: buildWindow(leftByDate),
    right: buildWindow(rightByDate),
  }
}

export function normalizeCumulativeReturn(window: ReturnWindow) {
  const anchorValue = window.points[0]?.value
  if (!(anchorValue > 0)) {
    return []
  }
  return window.points.map((point) => ({
    date: point.date,
    value: ((point.value / anchorValue) - 1) * 100,
  }))
}

export function cumulativeReturnPercentToGrowthIndex100(points: InstrumentChartPoint[]) {
  return points.map((point) => ({
    date: point.date,
    value: 100 + point.value,
  }))
}

export function commonObservationDateWindow(
  leftPoints: InstrumentChartPoint[],
  rightPoints: InstrumentChartPoint[],
) {
  const rightDates = new Set(
    normalizedReturnPoints(rightPoints).map((point) => point.date),
  )
  const commonDates = normalizedReturnPoints(leftPoints)
    .map((point) => point.date)
    .filter((pointDate) => rightDates.has(pointDate))
  return commonDates.length
    ? { start: commonDates[0], end: commonDates[commonDates.length - 1] }
    : null
}

export function periodReturnPercent(window: ReturnWindow) {
  const first = window.points[0]
  const last = window.points[window.points.length - 1]
  return first && last && first.value > 0 ? ((last.value / first.value) - 1) * 100 : null
}

export function actualYearFraction(startDate: string, endDate: string): number {
  if (endDate < startDate) {
    return -actualYearFraction(endDate, startDate)
  }
  const startYear = Number(startDate.slice(0, 4))
  const endYear = Number(endDate.slice(0, 4))
  const anniversary = (years: number) =>
    shiftIsoDate(startDate, { years }) || startDate
  let wholeYears = Math.max(endYear - startYear, 0)
  while (wholeYears > 0 && anniversary(wholeYears) > endDate) {
    wholeYears -= 1
  }
  const currentAnniversary = anniversary(wholeYears)
  if (currentAnniversary === endDate) {
    return wholeYears
  }
  const nextAnniversary = anniversary(wholeYears + 1)
  const current = new Date(`${currentAnniversary}T00:00:00Z`).getTime()
  const next = new Date(`${nextAnniversary}T00:00:00Z`).getTime()
  const end = new Date(`${endDate}T00:00:00Z`).getTime()
  return wholeYears + ((end - current) / (next - current))
}

export function annualizedReturnPercent(window: ReturnWindow) {
  const first = window.points[0]
  const last = window.points[window.points.length - 1]
  const start = new Date(`${window.anchorDate}T00:00:00Z`).getTime()
  const end = new Date(`${window.endDate}T00:00:00Z`).getTime()
  const days = Math.round((end - start) / 86_400_000)
  const years = actualYearFraction(window.anchorDate, window.endDate)
  if (
    !first ||
    !last ||
    first.value <= 0 ||
    days <= 0 ||
    years < 1
  ) {
    return null
  }
  return (Math.pow(last.value / first.value, 1 / years) - 1) * 100
}
